File: pixels-core/src/main/java/cn/edu/ruc/iir/pixels/core/reader/StringColumnReader.java
package cn.edu.ruc.iir.pixels.core.reader;

import cn.edu.ruc.iir.pixels.core.PixelsProto;
import cn.edu.ruc.iir.pixels.core.TypeDescription;
import cn.edu.ruc.iir.pixels.core.encoding.RunLenIntDecoder;
import cn.edu.ruc.iir.pixels.core.utils.BitUtils;
import cn.edu.ruc.iir.pixels.core.utils.DynamicIntArray;
import cn.edu.ruc.iir.pixels.core.vector.BytesColumnVector;
import cn.edu.ruc.iir.pixels.core.vector.ColumnVector;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufInputStream;
import io.netty.buffer.Unpooled;

import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * pixels
 *
 * @author guodong
 */
public class StringColumnReader
        extends ColumnReader
{
    private int originsOffset;
    private int startsOffset;
    private ByteBuf inputBuffer = null; // TODO: change ByteBuf to java.nio.ByteBuffer
    private ByteBuf originsBuf = null;
    private int[] orders = null;
    private int[] starts = null;
    private byte[] isNull = new byte[8];
    private ByteBuf contentBuf = null;
    private RunLenIntDecoder lensDecoder = null;
    private RunLenIntDecoder contentDecoder = null;
    private int isNullOffset = 0;
    private int isNullBitIndex = 0;

    StringColumnReader(TypeDescription type)
    {
        super(type);
    }

    /**
     * Read values from input buffer.
     *
     * @param input    input buffer
     * @param encoding encoding type
     * @param size     number of values to read
     * @param vector   vector to read into
     * @throws IOException
     */
    @Override
    public void read(ByteBuffer input, PixelsProto.ColumnEncoding encoding,
                     int offset, int size, int pixelStride, final int vectorIndex,
                     ColumnVector vector, PixelsProto.ColumnChunkIndex chunkIndex)
            throws IOException
    {
        BytesColumnVector columnVector = (BytesColumnVector) vector;
        if (offset == 0)
        {
            if (inputBuffer != null)
            {
                inputBuffer.release();
            }
            inputBuffer = Unpooled.wrappedBuffer(input);
            readContent(input.limit(), encoding);
            isNullOffset = (int) chunkIndex.getIsNullOffset();
            hasNull = true;
            elementIndex = 0;
            isNullBitIndex = 8;
        }
        // if dictionary encoded
        if (encoding.getKind().equals(PixelsProto.ColumnEncoding.Kind.DICTIONARY))
        {
            // read original bytes
            // we get bytes here to reduce memory copies and avoid creating many small byte arrays.
            byte[] buffer = originsBuf.array();
            // The available first byte in buffer should start from originsOffset.
            int bufferStart = originsOffset;
            for (int i = 0; i < size; i++)
            {
                if (elementIndex % pixelStride == 0)
                {
                    int pixelId = elementIndex / pixelStride;
                    hasNull = chunkIndex.getPixelStatistics(pixelId).getStatistic().getHasNull();
                    if (hasNull && isNullBitIndex > 0)
                    {
                        BitUtils.bitWiseDeCompact(isNull, inputBuffer.array(), isNullOffset++, 1);
                        isNullBitIndex = 0;
                    }
                }
                if (hasNull && isNullBitIndex >= 8)
                {
                    BitUtils.bitWiseDeCompact(isNull, inputBuffer.array(), isNullOffset++, 1);
                    isNullBitIndex = 0;
                }
                if (hasNull && isNull[isNullBitIndex] == 1)
                {
                    columnVector.isNull[i + vectorIndex] = true;
                    columnVector.noNulls = false;
                }
                else
                {
                    int originId = orders[(int) contentDecoder.next()];
                    int tmpLen;
                    if (originId < starts.length - 1)
                    {
                        tmpLen = starts[originId + 1] - starts[originId];
                    }
                    else
                    {
                        tmpLen = startsOffset - originsOffset - starts[originId];
                    }
                    // use setRef instead of setVal to reduce memory copy.
                    columnVector.setRef(i + vectorIndex, buffer, bufferStart + starts[originId], tmpLen);
                }
                if (hasNull)
                {
                    isNullBitIndex++;
                }
                elementIndex++;
            }
        }
        // if un-encoded
        else
        {
            // read values
            // we get bytes here to reduce memory copies and avoid creating many small byte arrays.
            byte[] buffer = contentBuf.array();
            int index = 0;
            for (int i = 0; i < size; i++)
            {
                if (elementIndex % pixelStride == 0)
                {
                    int pixelId = elementIndex / pixelStride;
                    hasNull = chunkIndex.getPixelStatistics(pixelId).getStatistic().getHasNull();
                    if (hasNull && isNullBitIndex > 0)
                    {
                        BitUtils.bitWiseDeCompact(isNull, inputBuffer.array(), isNullOffset++, 1);
                        isNullBitIndex = 0;
                    }
                }
                if (hasNull && isNullBitIndex >= 8)
                {
                    BitUtils.bitWiseDeCompact(isNull, inputBuffer.array(), isNullOffset++, 1);
                    isNullBitIndex = 0;
                }
                if (hasNull && isNull[isNullBitIndex] == 1)
                {
                    columnVector.isNull[i + vectorIndex] = true;
                    columnVector.noNulls = false;
                }
                else
                {
                    int len = (int) lensDecoder.next();
                    // use setRef instead of setVal to reduce memory copy.
                    columnVector.setRef(i + vectorIndex, buffer, index, len);
                    index += len;
                }
                if (hasNull)
                {
                    isNullBitIndex++;
                }
                elementIndex++;
            }
        }
    }

    private void readContent(int inputLength, PixelsProto.ColumnEncoding encoding)
            throws IOException
    {
        if (encoding.getKind().equals(PixelsProto.ColumnEncoding.Kind.DICTIONARY))
        {
            // read offsets
            inputBuffer.markReaderIndex();
            inputBuffer.skipBytes(inputLength - 3 * Integer.BYTES);
            originsOffset = inputBuffer.readInt();
            startsOffset = inputBuffer.readInt();
            int ordersOffset = inputBuffer.readInt();
            inputBuffer.resetReaderIndex();
            // read buffers
            contentBuf = inputBuffer.slice(0, originsOffset);
            originsBuf = inputBuffer.slice(originsOffset, startsOffset - originsOffset);
            ByteBuf startsBuf = inputBuffer.slice(startsOffset, ordersOffset - startsOffset);
            ByteBuf ordersBuf = inputBuffer.slice(ordersOffset, inputLength - ordersOffset);
            int originNum = 0;
            DynamicIntArray startsArray = new DynamicIntArray();
            RunLenIntDecoder startsDecoder = new RunLenIntDecoder(new ByteBufInputStream(startsBuf), false);
            while (startsDecoder.hasNext())
            {
                startsArray.add((int) startsDecoder.next());
                originNum++;
            }
            // read starts and orders
            RunLenIntDecoder ordersDecoder = new RunLenIntDecoder(new ByteBufInputStream(ordersBuf), false);
            starts = new int[originNum];
            orders = new int[originNum];
            for (int i = 0; i < originNum && ordersDecoder.hasNext(); i++)
            {
                starts[i] = startsArray.get(i);
                orders[i] = (int) ordersDecoder.next();
            }
            contentDecoder = new RunLenIntDecoder(new ByteBufInputStream(contentBuf), false);
        }
        else
        {
            // read lens field offset
            inputBuffer.markReaderIndex();
            inputBuffer.skipBytes(inputLength - Integer.BYTES);
            int lensOffset = inputBuffer.readInt();
            inputBuffer.resetReaderIndex();
            // read strings
            contentBuf = inputBuffer.slice(0, lensOffset);
            // read lens field
            ByteBuf lensBuf = inputBuffer.slice(lensOffset, inputLength - Integer.BYTES - lensOffset);
            lensDecoder = new RunLenIntDecoder(new ByteBufInputStream(lensBuf), false);
        }
    }
}

File: cpp/pixels-reader/pixels-core/lib/reader/StringColumnReader.cpp
//
// Created by liyu on 3/20/23.
//

#include "reader/StringColumnReader.h"
#include "profiler/CountProfiler.h"

StringColumnReader::StringColumnReader(std::shared_ptr<TypeDescription> type) : ColumnReader(type) {
    bufferOffset = 0;
	starts = nullptr;
}

void StringColumnReader::close() {

}

void StringColumnReader::read(std::shared_ptr<ByteBuffer> input, pixels::proto::ColumnEncoding & encoding, int offset,
                              int size, int pixelStride, int vectorIndex, std::shared_ptr<ColumnVector> vector,
                              pixels::proto::ColumnChunkIndex & chunkIndex, std::shared_ptr<pixelsFilterMask> filterMask) {
    if(offset == 0) {
        elementIndex = 0;
        bufferOffset = 0;
        readContent(input, input->bytesRemaining(), encoding);
    }
    // TODO: support dictionary
    std::shared_ptr<BinaryColumnVector> columnVector =
            std::static_pointer_cast<BinaryColumnVector>(vector);
    // TODO: if dictionary encoded
    if(filterMask != nullptr) {
        if (encoding.kind() == pixels::proto::ColumnEncoding_Kind_DICTIONARY) {
            for(int i = 0; i < size; i++) {
                if(elementIndex % pixelStride == 0) {
                    int pixelId = elementIndex / pixelStride;
                    // TODO: should write the remaining code
                }
                if(filterMask->get(i)) {
                    int originId = (int) contentDecoder->next();
                    int tmpLen = starts[originId + 1] - starts[originId];
                    // use setRef instead of setVal to reduce memory copy.
                    columnVector->setRef(i + vectorIndex, originsBuf->getPointer(), starts[originId], tmpLen);
                } else {
                    // skip this number
                    contentDecoder->next();
                }
                elementIndex++;
            }
        } else {
            for(int i = 0; i < size; i++) {
                if(elementIndex % pixelStride == 0) {
                    int pixelId = elementIndex / pixelStride;
                    // TODO: should write the remaining code
                }
                if(filterMask->get(i)) {
                    int len = (int) lensDecoder->next();
                    // use setRef instead of setVal to reduce memory copy
                    columnVector->setRef(
                            i + vectorIndex, contentBuf->getPointer(), bufferOffset, len);
                    bufferOffset += len;
                } else {
                    // skip this number
                    int len = (int) lensDecoder->next();
                    bufferOffset += len;
                }
                elementIndex++;
            }
        }
    } else {
        if (encoding.kind() == pixels::proto::ColumnEncoding_Kind_DICTIONARY) {
            for(int i = 0; i < size; i++) {
                if(elementIndex % pixelStride == 0) {
                    int pixelId = elementIndex / pixelStride;
                    // TODO: should write the remaining code
                }
                int originId = (int) contentDecoder->next();
                int tmpLen = starts[originId + 1] - starts[originId];
                // use setRef instead of setVal to reduce memory copy.
                columnVector->setRef(i + vectorIndex, originsBuf->getPointer(), starts[originId], tmpLen);
                elementIndex++;
            }
        } else {
            for(int i = 0; i < size; i++) {
                if(elementIndex % pixelStride == 0) {
                    int pixelId = elementIndex / pixelStride;
                    // TODO: should write the remaining code
                }
                int len = (int) lensDecoder->next();
                // use setRef instead of setVal to reduce memory copy
                columnVector->setRef(
                        i + vectorIndex, contentBuf->getPointer(), bufferOffset, len);
                bufferOffset += len;
                elementIndex++;
            }
        }
    }


}

void StringColumnReader::readContent(std::shared_ptr<ByteBuffer> input,
                                     uint32_t inputLength,
                                     pixels::proto::ColumnEncoding & encoding) {
    if(encoding.kind() == pixels::proto::ColumnEncoding_Kind_DICTIONARY) {
        input->markReaderIndex();
        input->skipBytes(inputLength - 2 * sizeof(int));
        originsOffset = input->getInt();
        startsOffset = input->getInt();
        input->resetReaderIndex();
        // read buffers
        contentBuf = std::make_shared<ByteBuffer>(*input, 0, originsOffset);
		originsBuf = std::make_shared<ByteBuffer>(
		    *input, originsOffset, startsOffset - originsOffset);
		std::shared_ptr<ByteBuffer> startsBuf = std::make_shared<ByteBuffer>(
		    *input, startsOffset, inputLength - startsOffset - 2 * sizeof(int));
		int bufferStart = 0;
		std::shared_ptr<RunLenIntDecoder> startsDecoder =
		    std::make_shared<RunLenIntDecoder>(startsBuf, false);
		if(encoding.has_dictionarysize()) {
			startsLength = encoding.dictionarysize() + 1;
			starts = new int[startsLength];
			int i = 0;
			while (startsDecoder->hasNext()) {
				starts[i++] = bufferStart + (int) startsDecoder->next();
			}
			starts[i] = bufferStart + startsOffset - originsOffset;
		} else {
            throw InvalidArgumentException("StringColumnReader::readContent: dictionary size must  be defined. ");
		}
		contentDecoder = std::make_shared<RunLenIntDecoder>(contentBuf, false);
    } else {
        input->markReaderIndex();
        input->skipBytes(inputLength - sizeof(int));
        int lensOffset = input->getInt();
        input->resetReaderIndex();
        // read strings
        // TODO: isDirect
        contentBuf = std::make_shared<ByteBuffer>(*input, 0, lensOffset);
        std::shared_ptr<ByteBuffer> lensBuf = std::make_shared<ByteBuffer>(
                *input, lensOffset, inputLength - sizeof(int) - lensOffset);
        lensDecoder = std::make_shared<RunLenIntDecoder>(lensBuf, false);
    }
}
StringColumnReader::~StringColumnReader() {
	if(starts != nullptr) {
		delete[] starts;
	}
}
